File: 2020/day/06/src/Main.hs
module Main where

import Data.Foldable (traverse_)
import Data.List (foldl1')
import Data.List.Split (splitOn)
import qualified Data.Set as Set
import System.Environment (getArgs)

main :: IO ()
main = (getArgs >>=) . traverse_ $ \inputFile -> do
  putStrLn inputFile

  input <- groups <$> readFile inputFile

  putStrLn $ "\tpart 1: sum over groups of number of questions that some member answered yes to"
  putStrLn $ "\t" ++ show (part1 input)

  putStrLn $ "\tpart 2: sum over groups of number of questions that all members answered yes to"
  putStrLn $ "\t" ++ show (part2 input)

type Group = [Member]

type Member = [Answer]

type Answer = Char

groups :: String -> [Group]
groups = map lines . splitOn "\n\n"

part1 :: [Group] -> Int
part1 = sum . map (Set.size . Set.unions . map Set.fromList)

part2 :: [Group] -> Int
part2 = sum . map (Set.size . foldl1' Set.intersection . map Set.fromList)

File: 2020/day/06/src/main.rs
use std::collections::HashSet;
use std::env;
use std::error;
use std::fs;

type Error = Box<dyn error::Error>;

fn main() -> Result<(), Error> {
    for input_file in env::args().skip(1) {
        println!("{}", input_file);
        let input = groups(&fs::read_to_string(input_file)?);

        println!(
            "\tpart 1: sum over groups of number of questions that some member answered yes to"
        );
        println!("\t{}", part1(&input));

        println!(
            "\tpart 2: sum over groups of number of questions that all members answered yes to"
        );
        println!("\t{}", part2(&input));
    }

    Ok(())
}

type Group = Vec<Member>;
type Member = HashSet<Answer>;
type Answer = char;

fn groups(s: &str) -> Vec<Group> {
    s.split("\n\n")
        .map(|s| s.lines().map(|line| line.chars().collect()).collect())
        .collect()
}

fn part1(gps: &Vec<Group>) -> usize {
    gps.iter()
        .map(|gp| {
            gp.iter()
                .fold(HashSet::new(), |acc, mem| acc.union(mem).cloned().collect())
                .len()
        })
        .sum()
}

fn part2(gps: &Vec<Group>) -> usize {
    gps.iter()
        .map(|gp| {
            gp.iter()
                .fold(None, |acc, mem| match acc {
                    None => Some(mem.clone()),
                    Some(acc) => Some(acc.intersection(mem).cloned().collect()),
                })
                .unwrap_or_else(HashSet::new)
                .len()
        })
        .sum()
}
